hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/AbstractEmbeddableMapping.java

```diff
@@ -108,9 +113,13 @@ public void setValues(Object component, Object[] values) {
 			optimizer.getAccessOptimizer().setPropertyValues( component, values );
 		}
 		else {
-			for ( int i = 0; i < values.length; i++ ) {
-				getAttributeMapping( i ).getPropertyAccess().getSetter().set( component, values[i] );
-			}
+			setAttributeValues( component, values );
```

Comment:
If we add the class name as last value, this method should check that the instance is of the correct type I guess.

Reply:
If we did this we would need to convert the discriminator value in `AbstractEmbeddableInitializer` to the class name though, which we avoided in favor of accessing the instantiator directly by object.

Reply:
I guess you're right and this method shouldn't do this check then to avoid the map lookup.